besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,369 @@
+
+    processTrieLogBatches(
+        rootWorldStateStorage,
+        blockchain,
+        chainHeight,
+        lastBlockNumberToRetainTrieLogsFor,
+        numberOfBatches,
+        batchFileNameBase);
+
+    if (rootWorldStateStorage.streamTrieLogKeys(layersToRetain).count() == layersToRetain) {
+      deleteFiles(batchFileNameBase, numberOfBatches);
+      LOG.info("Prune ran successfully. Enjoy some disk space back! \uD83D\uDE80");
+    } else {
+      LOG.error("Prune failed. Re-run the subcommand to load the trie logs from file.");
+    }
+  }
+
+  private static void processTrieLogBatches(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain,
+      final long chainHeight,
+      final long lastBlockNumberToRetainTrieLogsFor,
+      final long numberOfBatches,
+      final String batchFileNameBase) {
+
+    for (long batchNumber = 1; batchNumber <= numberOfBatches; batchNumber++) {
+      saveTrieLogBatches(
+          batchFileNameBase,
+          rootWorldStateStorage,
+          blockchain,
+          chainHeight,
+          lastBlockNumberToRetainTrieLogsFor,
+          batchNumber);
+    }
+
+    LOG.info("Clear trie logs...");
+    rootWorldStateStorage.clearTrieLog();
+
+    for (long batchNumber = 1; batchNumber <= numberOfBatches; batchNumber++) {
+      restoreTrieLogBatches(rootWorldStateStorage, batchNumber, batchFileNameBase);
+    }
+  }
+
+  private static void saveTrieLogBatches(
+      final String batchFileNameBase,
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain,
+      final long chainHeight,
+      final long lastBlockNumberToRetainTrieLogsFor,
+      final long batchNumber) {
+
+    final long firstBlockOfBatch = chainHeight - ((batchNumber - 1) * BATCH_SIZE);
+
+    final long lastBlockOfBatch =
+        Math.max(chainHeight - (batchNumber * BATCH_SIZE), lastBlockNumberToRetainTrieLogsFor);
+
+    final List<Hash> trieLogKeys = new ArrayList<>();
+
+    getTrieLogKeysForBlocks(blockchain, firstBlockOfBatch, lastBlockOfBatch, trieLogKeys);
+
+    LOG.info("Saving trie logs to retain in file (batch {})...", batchNumber);
+
+    try {
+      saveTrieLogsInFile(trieLogKeys, rootWorldStateStorage, batchNumber, batchFileNameBase);
+    } catch (IOException e) {
+      LOG.error("Error saving trie logs to file: {}", e.getMessage());
+    }
+  }
+
+  private static void restoreTrieLogBatches(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final long batchNumber,
+      final String batchFileNameBase) {
+
+    try {
+      LOG.info("Restoring trie logs retained from batch {}...", batchNumber);
+      recreateTrieLogs(rootWorldStateStorage, batchNumber, batchFileNameBase);
+    } catch (IOException | ClassNotFoundException e) {
+      LOG.error("Error recreating trie logs from batch {}: {}", batchNumber, e.getMessage());
```

Comment:
What should we do at this point? Do we have a way to restart from an existing restore file?

Reply:
I'm going to change this to throw a new Runtime exception so we abort the process here, meaning if you re-run the command it's gonna detect the files exist and skip the save to file part but it will still reset the db and try to restore from file.